db/compaction_picker.cc

```diff
@@ -1275,6 +1294,41 @@ bool LevelCompactionPicker::PickCompactionBySize(VersionStorageInfo* vstorage,
   return inputs->size() > 0;
 }
 
+bool LevelCompactionPicker::PickIntraL0Compaction(
+    VersionStorageInfo* vstorage, CompactionInputFiles* inputs) {
+  inputs->clear();
+  const std::vector<FileMetaData*>& level_files =
+      vstorage->LevelFiles(0 /* level */);
+  if (level_files.size() == 0 || level_files[0]->being_compacted) {
+    return false;
+  }
+
+  size_t compact_bytes = level_files[0]->fd.file_size;
+  size_t compact_bytes_per_del_file = port::kMaxSizet;
+  // compaction range will be [0, span_len).
+  size_t span_len;
+  // pull in files until the amount of compaction work per deleted file begins
+  // increasing.
+  for (span_len = 1; span_len < level_files.size(); ++span_len) {
+    compact_bytes += level_files[span_len]->fd.file_size;
+    size_t new_compact_bytes_per_del_file = compact_bytes / span_len;
+    if (level_files[span_len]->being_compacted ||
+        new_compact_bytes_per_del_file > compact_bytes_per_del_file) {
+      break;
+    }
+    compact_bytes_per_del_file = new_compact_bytes_per_del_file;
+  }
+
+  if (span_len >= kMinFilesForIntraL0Compaction) {
```

Comment:
Do we also want to introduce a minimum new_compact_bytes_per_del_file or something similar to avoid compacting big files (the files already compacted to be large through L0->L0)?

For example, we can say if new_compact_bytes_per_del_file is larger than write_buffer_size, or file_size_base_target, or level_size_base_target then we are going to skip it.